common/util/trees.h

```diff
@@ -0,0 +1,856 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This file contains generic functions and trait classes for operating on tree
+// structures.
+//
+// TreeNode concept
+// ----------------
+//
+// Class `T` must contain at least the following member in order to fulfill the
+// TreeNode concept:
+//
+// - `Container<T>& Children()`:
+//   Returns reference to a STL-like container containing the node's children.
+//
+// To check whether class `T` fulfills the concept one can use the
+// `TreeNodeTraits<T>` traits struct. It is defined only for classes fulfilling
+// the TreeNode concept, what makes it suitable for SFINAE tests. For
+// readability and consistency with Value and Parent checks (described below)
+// there is a member `TreeNodeTraits<T>::available` which is always true.
+//
+// Optional members:
+//
+// - `T* Parent()`:
+//   Returns pointer to a parent node or nullptr when the node is a tree root.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `ValueType& Value()`:
+//   Returns reference to a value stored in the node.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `subnodes_type` (typename):
+//   Container type which should be used for storing arrays of detached
+//   child nodes. It must be move-assignable to, and move-constructible from,
+//   a value returned by `Children()` method.
+//   When the type is absent, it is assumed to be the type returned by
+//   `Children()` method with reference and other modifiers removed.
+//   It is available through `TreeNodeTraits<T>::Children::container_type`.
+
+#ifndef VERIBLE_COMMON_UTIL_TREES_H_
+#define VERIBLE_COMMON_UTIL_TREES_H_
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iosfwd>  // IWYU pragma: keep
+#include <iterator>
+#include <numeric>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "common/util/logging.h"
+#include "common/util/spacer.h"
+#include "common/util/type_traits.h"
+
+namespace verible {
+
+namespace trees_internal {
+
+// TODO(mglb): move (Unavailable)FeatureTraits to a more generic header, or
+// replace them with something from standard library.
+
+// Type used as a placeholder for unavailable feature in type traits.
+struct UnavailableFeatureTraits {
+  static inline constexpr bool available = false;
+};
+
+// Type used as a base for available feature in type traits.
+struct FeatureTraits {
+  static inline constexpr bool available = true;
+};
+
+// TreeNodeTraits implementation details:
+
+// Alias to Node::subnodes_type if it exists.
+// Intended for use with detected_or_t.
+template <typename Node,  //
+          typename Type_ = typename Node::subnodes_type>
+using TreeNodeSubnodesType = Type_;
```

Comment:
should this type alias be hidden an a private or internal namespace?

Reply:
It is already hidden in `namespace types_internal`.